compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp

```diff
@@ -150,126 +151,114 @@ addTileAndDistributePasses(OpPassManager &funcPassManager,
+      for (auto [index, tileSize] :
+           llvm::enumerate(tilingLevelAttr.getSizes())) {
+        if (tileSize != 0 && pLoopsSet.contains(index)) {
+          return op->emitOpError(
+                     "expected only reduction dims to be set in the ")
+                 << IREE::CPU::getTilingLevelName(level)
+                 << " tiling level, got " << index << "-th tile size set";
+        }
       }
+      break;
     }
-  }
+    case IREE::CPU::TilingLevel::MaxNumTileLevels:
+    case IREE::CPU::TilingLevel::InvalidLevel:
+      break;
+    };
 
-  // Verify interchange.
-  for (int level = 0; level < tilingConfig.getNumTilingLevels(); level++) {
-    IREE::Codegen::LoweringConfigTilingLevelAttr attr =
-        tilingConfig.getTilingLevelAttr(level);
-    ArrayRef<int64_t> interchange = attr.getInterchange();
-    size_t expectedSize = attr.getSizes().size();
-    if (!interchange.empty() &&
-        !isValidInterchange(interchange, expectedSize)) {
+    ArrayRef<int64_t> interchange = tilingLevelAttr.getInterchange();
+    size_t expectedSize = tilingLevelAttr.getSizes().size();
+    if (!isValidInterchange(interchange, expectedSize)) {
       return op->emitOpError("expected [0, ")
-             << expectedSize << ") to be set exactly once in interchange #"
-             << level;
+             << expectedSize << ") to be set exactly once in interchange for "
+             << IREE::CPU::getTilingLevelName(level) << " tiling level";
     }
   }
+
   return success();
 }
 
 LogicalResult verifyConvTileAndDecomposeExpertConfig(
-    Operation *op, TilingConfig &tilingConfig,
-    IREE::Codegen::TranslationInfoAttr translationInfo,
-    ArrayRef<int64_t> workgroupSize) {
-  if (!isa<linalg::ConvolutionOpInterface>(op))
+    Operation *op, IREE::CPU::LoweringConfigAttr loweringConfig) {
+  if (!isa<linalg::ConvolutionOpInterface>(op)) {
     return success();
-
-  if (tilingConfig.getNumTilingLevels() == 6) {
-    // TODO: update verification.
-    return success();
-  }
-
-  if (tilingConfig.getNumTilingLevels() != 3) {
-    return op->emitOpError("expected three tiling levels, got ")
-           << tilingConfig.getNumTilingLevels();
   }
 
+  SmallVector<IREE::CPU::TilingLevel> requiredLevels = {
+      IREE::CPU::DistributionTiles, IREE::CPU::VectorCommonParallelTiles,
+      IREE::CPU::VectorReductionTiles};
   linalg::LinalgOp linalgOp = cast<linalg::LinalgOp>(op);
-  SmallVector<int64_t> shape = linalgOp.getStaticLoopRanges();
-  for (auto sizes : tilingConfig.getTileSizes()) {
-    for (auto [i, size] : llvm::enumerate(sizes)) {
-      if (size == 1)
-        shape[i] = 1;
-      if (shape[i] == -1 || size == 0)
+  SmallVector<int64_t> shapeAfterTiling = linalgOp.getStaticLoopRanges();
+  for (auto level : requiredLevels) {
+    if (!loweringConfig.hasTilingLevel(level)) {
+      return op->emitOpError("expected ")
+             << IREE::CPU::getTilingLevelName(level) << " is set";
+    }
+    for (auto [idx, size] :
```

Comment:
[nit] In the presence of "scalable" tile sizes, `size` will represent the "base size" (e.g. `4' in `vscale * 4 = [4]`). That's fine, but perhaps it's worth adding a comment?

Reply:
I see. I did not think of scalable tile sizes here. I'm going to mark the shape dynamic when the scalable flags are present.